secure-api-gateway-ig-extensions/src/test/java/com/forgerock/sapi/gateway/sign/DefaultSapiJwsSignerTest.java

```diff
@@ -0,0 +1,183 @@
+                "http://openbanking.org.uk/iat", System.currentTimeMillis() / 1000,
+                "http://openbanking.org.uk/iss", ASPSP_ORG_ID,
+                "http://openbanking.org.uk/tan", "openbanking.org.uk"
+        );
+    }
+
+    @Test
+    void shouldSignPayload() throws Exception {
+        // Given
+        DefaultSapiJwsSigner signPayload = new DefaultSapiJwsSigner(
+                getSecretsProvider(),
+                signingKeyId,
+                KID,
+                ALGORITHM
+        );
+        // When
+        String jwt = signPayload.critClaims(critClaims).sign(getPayloadMap());
+        // Then
+        assertThat(jwt).isNotNull();
+        final SignedJWT signedJwt = SignedJWT.parse(jwt);
+        validateSignature(signedJwt);
+        validateSignedJwt(signedJwt);
+    }
+
+    @Test
+    void shouldRaiseExceptionNullParameter() {
+        // Given / When / Then
+        assertThrows(
+                NullPointerException.class,
+                () -> new DefaultSapiJwsSigner(
+                        null,
+                        signingKeyId,
+                        KID,
+                        ALGORITHM
+                )
+        );
+    }
+
+    @Test
+    void shouldRaiseSignException() throws Exception {
+        // Given
+        String signingKeyId = "wrongKeyId";
+        SigningKey signingKey = getSigningKey(signingKeyId);
+
+        SecretsProvider secretsProvider = getSecretsProvider()
+                .useSpecificSecretForPurpose(
+                        Purpose.purpose("anotherKeyId", SigningKey.class),
+                        signingKey
+                );
+        DefaultSapiJwsSigner signPayload = new DefaultSapiJwsSigner(
+                secretsProvider,
+                signingKeyId,
+                KID,
+                ALGORITHM
+        );
+        // Then
+        assertThrows(Exception.class, () -> signPayload.critClaims(critClaims).sign(getPayloadMap()));
+    }
+
+    private SigningKey getSigningKey(String signingKeyId) throws Exception {
+        return new SecretBuilder().secretKey(keyPair.getPrivate())
+                .stableId(signingKeyId)
+                .keyUsages(singleton(SIGN))
+                .expiresAt(Instant.MAX)
+                .build(Purpose.SIGN);
+    }
+
+    private SecretsProvider getSecretsProvider() throws Exception {
+        SecretsProvider secretsProvider = new SecretsProvider(Clock.systemUTC());
+        secretsProvider.useSpecificSecretForPurpose(Purpose.purpose(signingKeyId, SigningKey.class),
+                getSigningKey(signingKeyId));
+        return secretsProvider;
+    }
+
+    private void validateSignature(SignedJWT signedJwt) {
+        try {
+            signedJwt.verify(rsaJwtVerifier);
+        } catch (JOSEException e) {
+            fail("Failed to verify signedJwt was signed by " + signingKeyId, e);
+        }
+    }
+
+    private void validateSignedJwt(SignedJWT signedJwt) throws ParseException {
```

Comment:
Add validation for the crit header

Reply:
The signature of sign method has been changed, the crit claims are optional and can be passed as null, depending the implementation could be validate to be mandatory and non null.
Critical claims validation added, and a new test to sign a payload with non critical claims.